### PR TITLE
Optional table parameter if query sql

### PR DESCRIPTION
**Pull request by HectorLagos**
Changing table to optional if sqlFile is present. https://github.com/spotify/dbeam/issues/127

### CODE REVIEW

File: dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java [hunk 1]
```diff
@@ -100,7 +100,7 @@ private String tsvMetrics() {
                 c ->
                     Optional.of(this.metrics.get(i).get(c))
                         .orElse(0L).toString()).collect(tabJoining),
-            this.metrics.get(i).get("bytesWritten") / this.metrics.get(i).get("writeElapsedMs")
+            this.metrics.get(i).get("bytesWritten")
```
Comment (labianchin) [inline dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java]: Can you elaborate on why this change?

File: dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java [hunk 2]
```diff
@@ -53,14 +70,27 @@ public void shouldFailParseOnInvalidArg() throws IOException, ClassNotFoundExcep
 
   @Test(expected = IllegalArgumentException.class)
   public void shouldFailOnMissingConnectionUrl() throws IOException, ClassNotFoundException {
-    optionsFromArgs("--table=somtable");
+    optionsFromArgs("--table=sometable");
   }
 
   @Test(expected = IllegalArgumentException.class)
-  public void shouldFailOnMissingTable() throws IOException, ClassNotFoundException {
+  public void shouldFailOnMissingTableSqlFile() throws IOException, ClassNotFoundException {
```
Comment (labianchin) [inline dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: Maybe `shouldFailOnMissingTableAndSqlFile`?

File: dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java [hunk 3]
```diff
@@ -53,14 +70,27 @@ public void shouldFailParseOnInvalidArg() throws IOException, ClassNotFoundExcep
...
+  public void shouldFailOnTableAndSqlFilePresent() throws IOException, ClassNotFoundException {
+    optionsFromArgs("--connectionUrl=jdbc:postgresql://some_db --sqlFile="
+        + sqlFile.getAbsolutePath() + " --table=some_table");
+  }
+
+  @Test
+  public void shouldNotFailOnMissingTableSqlFile() throws IOException, ClassNotFoundException {
+    JdbcExportArgs actual = optionsFromArgs("--connectionUrl=jdbc:postgresql://some_db --sqlFile="
+        + sqlFile.getAbsolutePath());
+  }
```
Comment (labianchin) [inline dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: Consider having a few asserts on `actual` contents like `table` and `sqlFile`.

File: dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java [hunk 4]
```diff
@@ -117,9 +117,11 @@ public static QueryBuilderArgs createQueryArgs(JdbcExportPipelineOptions options
 
   private static QueryBuilderArgs createQueryBuilderArgs(JdbcExportPipelineOptions options)
       throws IOException {
+    checkArgument((options.getTable() != null) != (options.getSqlFile() != null),
+            "Either --table or --sqlFile must be present");
     if (options.getSqlFile() != null) {
       return QueryBuilderArgs.create(
-          options.getTable(), PasswordReader.readFromFile(options.getSqlFile()));
+          "user_based_query", PasswordReader.readFromFile(options.getSqlFile()));
```
Comment (labianchin) [inline dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java]: Is there a ways users can configure `"user_based_query"` to something else?

Another thought, maybe `table` name is not important. We could consider having two different implementations of `QueryBuilderArgs`: one "table name" based and another `sqlFile` based. Not that we necessary need to solve it in this PR, just sharing a thought and asking for your thoughts.

File: dbeam-core/src/test/java/com/spotify/dbeam/jobs/JdbcAvroJobTest.java [hunk 5]
```diff
@@ -124,6 +129,45 @@ public void shouldRunJdbcAvroJob() throws IOException {
     Assert.assertEquals(2, records.size());
   }
 
+  @Test
+  public void shouldRunJdbcAvroJobSqlFile() throws IOException {
```
Comment (labianchin) [inline dbeam-core/src/test/java/com/spotify/dbeam/jobs/JdbcAvroJobTest.java]: Nice to have this new test! Great job!

**Comment by labianchin** [comment]
Overall looks good! Left a few comments and I would like @ra1861 to also take a look, since he worked on this feature before.

File: dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java (hunk 1 above)
Comment (HectorLagos) [reply dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java]: for some reason (sometimes), executing the tests locally, the bench test is throwing an ArithmeticException: /by zero. have you seen this behavior ?if it is not intended to run locally i can revert the changes and skip this test

File: dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java (hunk 2 above)
Comment (HectorLagos) [reply dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: agree

File: dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java (hunk 4 above)
Comment (HectorLagos) [reply dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java]: That was the original idea! however I  found table name used in several places for several purposes (part of JobConfiguration, QueryBuilder and  schema generation). I think it would be useful to create more explicit parameters for this purposes. We may want to add options for each and used tablename as default. can you create a story for it ? I can help with it

Comment (ra1861) [reply dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java]: I would go even further and remove:

- QueryBuilderArgs.tableName()
- QueryBuilderArgs.Builder.setTableName()

and remove _tableName_ parameter from:

- QueryBuilderArgs.create(String tableName, String sqlQueryOpt)

File: dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java (hunk 1 above)
Comment (ra1861) [reply dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java]: This seems to be unrelated to the main change. Might be put into a separate PR.

File: dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java [hunk 6]
```diff
@@ -23,18 +23,36 @@
...
+import org.junit.BeforeClass;
 import org.junit.Test;
 
 
 public class JdbcExportOptionsTest {
+  private static File sqlFile;
+
+  @BeforeClass
+  public static void beforeAll() throws IOException {
+    sqlFile = File.createTempFile("query", ".sql");
```
Comment (ra1861) [inline dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: I just wonder why there are two different approaches to create temp files.
>sqlFile = File.createTempFile("query", ".sql");
vs
>private static File SQL_FILE = new File(DIR.getAbsolutePath() + ".sql");

Comment (HectorLagos) [reply dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: I found the same between PASSWORD_FILE in JdbcAvroJobTest and  passwordFile in PasswordReaderTest. Better to have one single way, should we  keep create File.createTempFile ?

Comment (ra1861) [reply dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: I don't really have an opinion which one is better. I think it is nice to select one and refactor it to a separate method in some test helper class.

Comment (HectorLagos) [reply dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: agree, do you want to do it in the same PR ?

File: dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java (hunk 4 above)
Comment (HectorLagos) [reply dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java]: I see few options to get rid off tableName.

- keep  setTableName and remove getTableName(), at the time QueryBuilderArgs creation if tablename is present, baseSqlQuery can be set based on tablename. Then queryBuilder would work only on top of baseSqlQuery rather than 2 different ways. Cons: passing additional logic to QueryBuilderArgs creation to build  base sql
- remove setTableName and getTableName from QueryBuilderArgs. Then we will need to pass additional table param to QueryBuilder from DBeamPipelineOptions. Cons: I think QueryBuilderArgs lose its purpose if we need inject additional params to QueryBuilder outside of this
- Remove support for tablename and force sqlFile usage. I would not go for this option, I think it is a nice feature to have this. 

Let me know what you think

File: dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java (hunk 1 above)
Comment (labianchin) [reply dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java]: I see. Then maybe `if this.metrics.get(i).get("writeElapsedMs") == 0` then it should be 0. Otherwise just to the division.

Feel free to have this as a separate PR.

File: dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java (hunk 6 above)
Comment (labianchin) [reply dbeam-core/src/test/java/com/spotify/dbeam/args/JdbcExportOptionsTest.java]: I think `File.createTempFile` feels nice!

I don't think it is necessary to address it in this PR. Though it is something for us to follow up.

File: dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java (hunk 4 above)
Comment (labianchin) [reply dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java]: 1. Exporting `--table` parameter is an important use case. Personally I think it is more important than `--sqlFile`. Regardless, I think there is space for both on DBeam.
2. When breaking APIs (e.g. removing a parameter), we would need to have a good justification to introduce such friction on DBeam users. I don't think it is the case here.
3. I think `QueryBuilderArgs` needs some refactoring, it is the only (non builder) class that is mutable. I would like to see that part immutable and with a better logic for both tableName and sqlFile.

I think don't yet see a `QueryBuilderArgs` without `tableName` parameter. Maybe it is possible, maybe worth to try.

Comment (ra1861) [reply dbeam-core/src/main/java/com/spotify/dbeam/options/JdbcExportArgsFactory.java]: The code LGTM, we can address this and continue discussion of improvements in a separate PR.

File: dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java (hunk 1 above)
Comment (HectorLagos) [reply dbeam-core/src/main/java/com/spotify/dbeam/jobs/BenchJdbcAvroJob.java]: Will take this change out and create another PR

**Comment by HectorLagos** [comment]
@labianchin @ra1861 should be ready with what discussed, will address division by zero in benchmark, create testhelper method for creating temp file and refactor for QueryBuilderArgs independently
